### PR TITLE
Add a refresh command that rebuilds the snapshot and reloads it on th…

### PR DESCRIPTION
**Pull request by jason-simmons**
…e device

This provides a fast way to iterate on changes to a Flutter app that only
involve updates to Dart code and do not require a full build and install
of the FLX and APK

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/refresh.dart [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      printError('No device connected.');
+      return 1;
+    }
+
+    Directory tempDir = await Directory.systemTemp.createTemp('flutter_tools');
+    try {
+      String snapshotPath = path.join(tempDir.path, 'snapshot_blob.bin');
+
+      int result = await toolchain.compiler.compile(
+          mainPath: argResults['target'], snapshotPath: snapshotPath);
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/refresh.dart]: if you have a newline after ( then have a newline before )

File: packages/flutter_tools/lib/src/commands/refresh.dart [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      if (result != 0) {
+        printError('Failed to run the Flutter compiler. Exit code: $result');
+        return result;
+      }
+
+      bool success = await devices.android.refreshSnapshot(
+          applicationPackages.android, snapshotPath);
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/refresh.dart]: ditto

File: packages/flutter_tools/lib/src/commands/refresh.dart [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+import 'package:path/path.dart' as path;
+
+import '../base/context.dart';
+import '../flx.dart';
+import '../runner/flutter_command.dart';
+
+class RefreshCommand extends FlutterCommand {
+  final String name = 'refresh';
+  final String description = 'Reload the Dart code in a Flutter app (Android only).';
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/refresh.dart]: Make it clear that this involves rebuilding the FLX, i.e. that host local changes take effect, as opposed to just restarting the app with the same code as was used before.

**Comment by Hixie** [comment]
Can we do this automatically somehow? e.g. track the timestamps on our dependencies and if nothing changed, just do this instead?

LGTM

File: packages/flutter_tools/lib/src/commands/refresh.dart (hunk 1 above)
Comment (jason-simmons) [inline packages/flutter_tools/lib/src/commands/refresh.dart]: done

File: packages/flutter_tools/lib/src/commands/refresh.dart (hunk 2 above)
Comment (jason-simmons) [inline packages/flutter_tools/lib/src/commands/refresh.dart]: done

File: packages/flutter_tools/lib/src/commands/refresh.dart (hunk 3 above)
Comment (jason-simmons) [inline packages/flutter_tools/lib/src/commands/refresh.dart]: done

**Comment by jason-simmons** [comment]
Eventually the flutter command will probably adopt the features of a full-fledged build system, but today we don't have robust dependency tracking.

This command is for the common scenario where you have run flutter start and deployed an APK to the target device, and you want to roll out new Dart code but haven't changed anything else in the package (resources, services, etc.)
In particular, this avoids the cost of building a full FLX and copying it to the device.

**Pull request merged**